### PR TITLE
Tile container

### PR DESCRIPTION
**Pull request by Elv13**
Long time ago, I told @actionless I would add a new `awful.wallpaper` module to be able to use `svg` and normal widgets to create a wallpaper. It's kind of blocking 4.4 "because I told people I would do it". There was 2 blocker PRs, one to improve `wibox.widget.imagebox` to add "stretch" properties. The other was a `tile` widget to make tiled wallpaper. I finally finished this one.

@actionless Since you are the only person I know who use tiled (Pokemon!) wallpapers. Do you mind doing a review on this? Thanks.

image

### CODE REVIEW

**Comment by actionless** [comment]
speaking of wallpaper, mb is there some quick route to also fix this one? https://github.com/awesomeWM/awesome/pull/2867#issuecomment-538418756

**Comment by Elv13** [comment]
> speaking of wallpaper, mb is there some quick route to also fix this one? #2867 (comment)

As mentionned in the first message, supporting SVG is the main goal of `awful.wallpaper`. Right now the SVG rendering logic would be impractical to bold on to `gears.wallpaper`. There is some non-trivial corner cases which are currently handled correctly in `wibox.container.imagebox`. I would like the wallpaper code to re-use that imagebox code. But first we need to bring imagebox/tile to feature parity with the wallpaper API.

This PR is about getting closer to that.

File: lib/wibox/widget/slider.lua [hunk 1]
```diff
@@ -381,7 +381,7 @@ function slider:draw(_, cr, width, height)
 
     if bar_active_color and type(bar_color) == "string" and type(bar_active_color) == "string" then
         local bar_active_width = active_rate * (width - x_offset - right_margin)
-            - (handle_width - handle_border_width/2) * (active_rate - 0.5)
+            - math.floor(handle_width - handle_border_width/2) * (active_rate - 0.5)
```
Comment (actionless) [inline lib/wibox/widget/slider.lua:384]: why here and next you making floor() not of whole equation but only to some part?

**Review by Aire-One — COMMENTED** [review]
Looks good to me overall. 

The tile drawing seems a little blurry on my end, tho. I tried to investigate this for a bit, and it seems it has something to do with the EXTEND_REPEAT Pattern https://lists.cairographics.org/archives/cairo/2012-January/022646.html
(https://stackoverflow.com/questions/54868972/how-can-i-prevent-cairo-from-rasterizing-my-pattern-fills is a more recent example)

File: lib/wibox/container/tile.lua [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+local defaults = {
+    horizontal_spacing = 0,
+    vertical_spacing   = 0,
+    tiled              = true,
+    horizontal_crop    = false,
+    vertical_crop      = false,
+}
+
+for _, prop in ipairs {"horizontal_spacing", "vertical_spacing", "tiled",
```
[the remark this answers is not in the document]
Comment (Elv13) [reply lib/wibox/container/tile.lua]: done

File: lib/wibox/container/tile.lua [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+-- @tparam wibox.widget widget args.widget The widget to tile.
+-- @tparam string args.halign Either `left`, `right` or `center`.
+-- @tparam string args.valign Either `top`, `bottom` or `center`.
+-- @constructorfct wibox.container.tile
+function module.mt:__call(...)
+    return new(...)
+end
+
+return setmetatable(module, module.mt)
+
```
[the remark this answers is not in the document]
Comment (Elv13) [reply lib/wibox/container/tile.lua:210]: done

File: lib/wibox/widget/slider.lua (hunk 1 above)
Comment (Elv13) [reply lib/wibox/widget/slider.lua]: done

**Comment by Elv13** [comment]
> The tile drawing seems a little blurry on my end, tho. I tried to investigate this for a bit, and it seems it has something to do with the EXTEND_REPEAT Pattern https://lists.cairographics.org/archives/cairo/2012-January/022646.html
(https://stackoverflow.com/questions/54868972/how-can-i-prevent-cairo-from-rasterizing-my-pattern-fills is a more recent example)

This is mostly only in the doc because it rasterize them and they are displayed with a different resolution. In practice, it seems fine. Did you notice runtime glitches?

**Comment by Aire-One** [comment]
> > The tile drawing seems a little blurry on my end, tho. I tried to investigate this for a bit, and it seems it has something to do with the EXTEND_REPEAT Pattern https://lists.cairographics.org/archives/cairo/2012-January/022646.html
> > (https://stackoverflow.com/questions/54868972/how-can-i-prevent-cairo-from-rasterizing-my-pattern-fills is a more recent example)
> 
> This is mostly only in the doc because it rasterize them and they are displayed with a different resolution. In practice, it seems fine. Did you notice runtime glitches?

Yeap, you're right. I should have started by trying to use it instead of focusing on generated images for the doc...

**Review by Aire-One — APPROVED** [review]
LGTM :)

**Review by actionless — APPROVED** [review]

**Pull request merged**